argon2/src/instance.rs

```diff
@@ -115,10 +122,54 @@ impl<'a> Instance<'a> {
         Ok(instance)
     }
 
+    /// Create multiple mutable references for the current instance, one for every lane
+    #[cfg(feature = "parallel")]
+    #[allow(unsafe_code)]
+    unsafe fn mut_self_refs(&mut self) -> Vec<&'static mut Instance<'static>> {
+        let lanes = self.lanes;
+        // This transmute can be skipped when a scoped threadpool is used (or when `spawn_unchecked()` gets stabilised)
+        let this = mem::transmute::<_, &mut Instance<'static>>(self);
+        let this: *mut Instance<'static> = this;
+
+        // Dereference the raw pointer multiple times to create multiple mutable references
+        (0..lanes).map(|_| &mut *(this)).collect()
+    }
+
+    #[cfg(feature = "parallel")]
+    fn fill_memory_blocks_par(&mut self) {
+        for r in 0..self.passes {
+            for s in 0..SYNC_POINTS {
+                // Safety: - All threads that receive a references will be joined before the item gets dropped
+                //         - All the read and write operations *shouldn't* overlap
+                #[allow(unsafe_code)]
+                let self_refs = unsafe { self.mut_self_refs() };
+
+                (0..self.lanes)
+                    .zip(self_refs)
+                    .par_bridge()
+                    .for_each(|(l, self_ref)| {
+                        self_ref.fill_segment(Position {
```

Comment:
What does `fill_segment` do?

Reply:
This is hopefully answered in my comments below, but it's effectively the minimum unit of work that a single worker thread can operate on at a given time.

It pseudorandomly samples reference inputs from other "slices" (in the Argon2 nomeclature) that aren't currently being worked on by other threads, and uses them to fill a particular "segment" with pseudorandom values (computed using a hash function).